File: backend/src/main/java/org/akashihi/mdg/dao/TransactionRepositoryCustomImpl.java
package org.akashihi.mdg.dao;

import lombok.RequiredArgsConstructor;
import org.akashihi.mdg.entity.Transaction;
import org.springframework.data.jpa.domain.Specification;

import javax.persistence.EntityManager;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import java.util.stream.Stream;

@RequiredArgsConstructor
public class TransactionRepositoryCustomImpl implements TransactionRepositoryCustom {
    private final EntityManager em;

    @Override
    public Stream<Transaction> streamByAccount(Specification<Transaction> spec) {
        CriteriaBuilder cb = em.getCriteriaBuilder();
        CriteriaQuery<Transaction> query = cb.createQuery(Transaction.class);
        Root<Transaction> root = query.from(Transaction.class);
        query.where(spec.toPredicate(root, query, cb));

        return em.createQuery(query).getResultStream();
    }
}

File: backend/src/main/java/org/akashihi/mdg/dao/TransactionRepositoryCustomImpl.kt
package org.akashihi.mdg.dao

import org.akashihi.mdg.entity.Transaction
import org.springframework.data.jpa.domain.Specification
import java.util.stream.Stream
import javax.persistence.EntityManager

class TransactionRepositoryCustomImpl(private val em: EntityManager) : TransactionRepositoryCustom {
    override fun streamByAccount(spec: Specification<Transaction>): Stream<Transaction> {
        val cb = em.criteriaBuilder
        val query = cb.createQuery(Transaction::class.java)
        val root = query.from(Transaction::class.java)
        query.where(spec.toPredicate(root, query, cb))
        return em.createQuery(query).resultStream
    }
}